### PR TITLE
[release-0.3] Fix/improve Travis cache handling  [skip appveyor]

### PR DESCRIPTION
**Pull request by blueyed**
Via https://github.com/neovim/neovim/pull/10412.

### CODE REVIEW

**Comment by blueyed** [comment]
Cache was busted as expected: https://travis-ci.org/neovim/neovim/jobs/554339377#L545

Please review the yaml magic extra carefully though, i.e. if jobs are generated correctly.
Also the `allow_failures` was not tested (but maybe will here due to a flaky test).

File: .travis.yml [hunk 1]
```diff
@@ -46,47 +46,71 @@ env:
...
       # Use Lua so that ASAN can test our embedded Lua support. 8fec4d53d0f6
-      env: >
-        CLANG_SANITIZER=ASAN_UBSAN
-        CMAKE_FLAGS="$CMAKE_FLAGS -DPREFER_LUA=ON"
-        ASAN_SYMBOLIZE=asan_symbolize-4.0
+      env:
+        <<: *common-job-env
+        CLANG_SANITIZER: ASAN_UBSAN
+        CMAKE_FLAGS: "$CMAKE_FLAGS -DPREFER_LUA=ON"
+        ASAN_SYMBOLIZE: asan_symbolize-4.0
```
Comment (justinmk) [inline .travis.yml]: why was `=` change to `:` ?

**Comment by justinmk** [comment]
> Please review the yaml magic extra carefully though

it's kind of impossible to feel confident about YAML :)

File: .travis.yml [hunk 2]
```diff
@@ -46,47 +46,71 @@ env:
...
     - os: linux
       compiler: gcc
-      env: >
-        FUNCTIONALTEST=functionaltest-lua
-        CMAKE_FLAGS="$CMAKE_FLAGS -DPREFER_LUA=ON"
-        DEPS_CMAKE_FLAGS="$DEPS_CMAKE_FLAGS -DUSE_BUNDLED_LUAJIT=OFF"
+      env:
+        <<: *common-job-env
```
Comment (justinmk) [inline .travis.yml]: better if it goes at the end, since it is just noise in the travis UI.

File: .travis.yml (hunk 1 above)
Comment (blueyed) [reply .travis.yml]: Because it became a hash.  Now it's a list (to have it in the end).

File: .travis.yml (hunk 2 above)
Comment (blueyed) [reply .travis.yml]: ok.

**Comment by blueyed** [comment]
btw: it is very unfortunate actually that job env vars result in different caches - it would be good if this could be controlled (https://travis-ci.community/t/using-unified-cache-control-cache-identity/1531).

**Comment by blueyed** [comment]
We can use names: https://travis-ci.org/neovim/neovim/builds/554512328
envs still show up with jobs then: https://travis-ci.org/neovim/neovim/jobs/554512329 (but it is at the end there now, and descriptive)

**Comment by blueyed** [comment]
allow_failures with names worked: https://travis-ci.org/neovim/neovim/builds/554514583

Reverting tests, should be good.

**Comment by blueyed** [comment]
We can have the names to be longer / more descriptive - suggestions welcome, of course.

**Comment by justinmk** [comment]
Beautiful!

File: .travis.yml [hunk 3]
```diff
@@ -47,46 +47,74 @@ env:
...
+      env:
+        - BUILD_32BIT: ON
+        - *common-job-env
+    - name: "OSX: clang"
+      os: osx
       compiler: clang
       osx_image: xcode9.4  # macOS 10.13
-    - os: osx
+      env: *common-job-env
+    - name: "OSX: gcc"
```
Comment (justinmk) [inline .travis.yml]: Let's call it "macOS" instead of "OSX", on master :)

File: .travis.yml [hunk 4]
```diff
@@ -47,46 +47,74 @@ env:
...
+        - *common-job-env
+    - name: gcc-32bit
+      os: linux
       # Travis creates a cache per compiler. Set a different value here to
       # store 32-bit dependencies in a separate cache.
       compiler: gcc
-      env: BUILD_32BIT=ON
-    - os: osx
+      env:
+        - BUILD_32BIT: ON
```
Comment (blueyed) [inline .travis.yml]: Should have been BUILD_32BIT=ON - but I guess Travis handles this.

**Pull request merged**